gemrb/core/System/OS4Helpers.cpp

```diff
@@ -0,0 +1,730 @@
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <math.h>
+#include <cassert>
+#include <unistd.h>
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+/* This code to undef const added in libiberty.  */
+#ifndef __STDC__
+/* This is a separate conditional since some stdc systems
+   reject `defined (const)'.  */
+#ifndef const
+#define const
+#endif
+#endif
+
+#include "OS4Helpers.h"
+
+namespace std {
+
+    int iswspace(wint_t c) {
+#ifdef _MB_CAPABLE
+          //c = _jp2uc (c);
+        return ((c >= 0x0009 && c <= 0x000d) || c == 0x0020 ||
+          c == 0x1680 || c == 0x180e ||
+          (c >= 0x2000 && c <= 0x2006) ||
+          (c >= 0x2008 && c <= 0x200a) ||
+          c == 0x2028 || c == 0x2029 ||
+          c == 0x205f || c == 0x3000);
+#else
+        return (c < 0x100 ? isspace (c) : 0);
+#endif /* _MB_CAPABLE */
+    }
+
+    long int lround(double x) {
+        return round(x);
+    }
+}
+
+static int _putwc_cb(wchar_t wc, void *abstract) {
+	_putwc_data_cb *data = (_putwc_data_cb*) abstract;
+	char buf[8];
+	size_t len;
+
+	len = wcrtomb(buf, wc, &data->mbs);
+	if (len != (size_t)-1) {
+		if (fwrite(buf, 1, len, data->fp) == len)
+			return 0;
+	}
+
+	return -1;
+}
+
+int vfwprintf(FILE *fp, const wchar_t *fmt, va_list ap) {
+	_putwc_data_cb cb_data;
+
+	cb_data.fp = fp;
+	memset(&cb_data.mbs, 0, sizeof(cb_data.mbs));
+
+	return _libwprintf_dofmt(_putwc_cb, &cb_data, fmt, ap);
+}
+
+
+int swprintf(wchar_t *buffer, size_t maxlen, const wchar_t *fmt, ...) {
+	va_list ap;
+	int wc;
+
+	va_start(ap, fmt);
+	wc = vswprintf(buffer, maxlen, fmt, ap);
+	va_end(ap);
+
+	return wc;
+}
+
+int vswprintf(wchar_t *buffer, size_t maxlen, const wchar_t *fmt, va_list ap) {
+	_putwc_data cb_data;
+	int result;
+
+	cb_data.buffer = buffer;
+	cb_data.space  = maxlen;
+
+	result = _libwprintf_dofmt(_putwc_cb, &cb_data, fmt, ap);
+
+	/* Make sure that output is NUL terminated */
+	_putwc_cb('\0', &cb_data);
+
+	return result;
+}
+
+#define PUTWC(wc)                   \
+	__extension__ ({                              \
+		if (cb((wc), cb_data) == 0) \
+			count++;                \
+		else                        \
+			return -1;              \
+	})
+
+static void reverse(char *str, size_t len) {
+	if (len > 1) {
+		char *start = str;
+		char *end = str + len - 1;
+		char tmp;
```

Comment:
Codacy found an issue: [The scope of the variable 'tmp' can be reduced.](https://app.codacy.com/manual/gemrb/gemrb/pullRequest?prid=6340351)